src/node/routes/vscode.ts

```diff
@@ -1,65 +1,105 @@
-}
-
-export const createVSServerRouter = async (args: DefaultedArgs): Promise<VSServerResult> => {
-  // See ../../../vendor/modules/code-oss-dev/src/vs/server/main.js.
-  const createVSServer = await loadAMDModule<CodeServerLib.CreateServer>(
-    "vs/server/remoteExtensionHostAgent",
-    "createServer",
-  )
+export class CodeServerRouteWrapper {
+  /** Assigned in `ensureCodeServerLoaded` */
+  private _codeServerMain!: CodeServerLib.IServerAPI
+  private _wsRouterWrapper = WsRouter()
+  public router = express.Router()
 
-  const codeServerMain = await createVSServer(null, {
-    connectionToken: "0000",
-    ...args,
-    // For some reason VS Code takes the port as a string.
-    port: typeof args.port !== "undefined" ? args.port.toString() : undefined,
-  })
+  public get wsRouter() {
+    return this._wsRouterWrapper.router
+  }
 
-  const router = express.Router()
-  const wsRouter = WsRouter()
+  //#region Route Handlers
 
-  router.get("/", async (req, res, next) => {
+  private $root: express.Handler = async (req, res, next) => {
     const isAuthenticated = await authenticated(req)
+
     if (!isAuthenticated) {
       return redirect(req, res, "login", {
         // req.baseUrl can be blank if already at the root.
         to: req.baseUrl && req.baseUrl !== "/" ? req.baseUrl : undefined,
       })
     }
+
     next()
-  })
+  }
 
-  router.all("*", ensureAuthenticated, (req, res, next) => {
-    req.on("error", (error: any) => {
+  private $proxyRequest: express.Handler = async (req, res, next) => {
+    // We allow certain errors to propagate so that other routers may handle requests
+    // outside VS Code
+    const requestErrorHandler = (error: any) => {
       if (error instanceof Error && ["EntryNotFound", "FileNotFound", "HttpError"].includes(error.message)) {
         next()
       }
 
       errorHandler(error, req, res, next)
-    })
+    }
+
+    req.once("error", requestErrorHandler)
 
-    codeServerMain.handleRequest(req, res)
-  })
+    this._codeServerMain.handleRequest(req, res)
+  }
 
-  wsRouter.ws("/", ensureAuthenticated, (req) => {
-    codeServerMain.handleUpgrade(req, req.socket)
+  private $proxyWebsocket = async (req: WebsocketRequest) => {
+    this._codeServerMain.handleUpgrade(req, req.socket)
 
     req.socket.resume()
-  })
+  }
+
+  //#endregion
+
+  /**
+   * Fetches a code server instance asynchronously to avoid an initial memory overhead.
+   */
+  private ensureCodeServerLoaded: express.Handler = async (req, _res, next) => {
+    if (this._codeServerMain) {
+      return next()
+    }
+
+    const { args } = req
+
+    /**
+     * @file ../../../vendor/modules/code-oss-dev/src/vs/server/main.js
+     */
+    const createVSServer = await loadAMDModule<CodeServerLib.CreateServer>(
+      "vs/server/remoteExtensionHostAgent",
+      "createServer",
+    )
+
+    try {
+      this._codeServerMain = await createVSServer(null, {
+        connectionToken: "0000",
```

Comment:
❓ What's the `connectionToken` used for again?

Reply:
Connection token is used by VS Code as a basic form of client-side security. At the moment we’ve disabled its usage, but I think it’s worth looking at again as upstream matures the feature.

Reply:
Related: https://github.com/cdr/code-server/issues/4479